src/pod.rs

```diff
@@ -60,6 +56,216 @@ pub(crate) fn bytes_of<T: Pod>(val: &T) -> &[u8] {
     unsafe { slice::from_raw_parts(slice::from_ref(val).as_ptr().cast(), size) }
 }
 
+/// A newtype for byte slices.
+///
+/// It has these important features:
+/// - no methods that can panic, such as `Index`
+/// - convenience methods for `Pod` types
+/// - a useful `Debug` implementation
+#[derive(Default, Clone, Copy, PartialEq, Eq)]
+pub struct Bytes<'data>(pub &'data [u8]);
+
+impl<'data> fmt::Debug for Bytes<'data> {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        debug_list_bytes(self.0, fmt)
+    }
+}
+
+impl<'data> Bytes<'data> {
+    #[inline]
+    pub fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    #[inline]
+    pub fn skip(&mut self, offset: usize) -> Option<()> {
+        match self.0.get(offset..) {
+            Some(tail) => {
+                self.0 = tail;
+                Some(())
+            }
+            None => {
+                self.0 = &[];
+                None
+            }
+        }
+    }
+
+    #[inline]
+    pub fn read_bytes(&mut self, count: usize) -> Option<Bytes<'data>> {
+        match (self.0.get(..count), self.0.get(count..)) {
+            (Some(head), Some(tail)) => {
+                self.0 = tail;
+                Some(Bytes(head))
+            }
+            _ => {
+                self.0 = &[];
+                None
+            }
+        }
+    }
+
+    #[inline]
+    pub fn read_bytes_at(mut self, offset: usize, count: usize) -> Option<Bytes<'data>> {
+        self.skip(offset)?;
+        self.read_bytes(count)
+    }
+
+    #[inline]
+    pub fn read<T: Pod>(&mut self) -> Option<&'data T> {
+        match from_bytes(self.0) {
+            Some((value, tail)) => {
+                self.0 = tail;
+                Some(value)
+            }
+            None => {
+                self.0 = &[];
+                None
+            }
+        }
+    }
+
+    #[inline]
+    pub fn read_at<T: Pod>(mut self, offset: usize) -> Option<&'data T> {
+        self.skip(offset)?;
+        self.read()
+    }
+
+    #[inline]
+    pub fn read_slice<T: Pod>(&mut self, count: usize) -> Option<&'data [T]> {
+        match slice_from_bytes(self.0, count) {
+            Some((value, tail)) => {
+                self.0 = tail;
+                Some(value)
+            }
+            None => {
+                self.0 = &[];
+                None
+            }
+        }
+    }
+
+    #[inline]
+    pub fn read_slice_at<T: Pod>(mut self, offset: usize, count: usize) -> Option<&'data [T]> {
+        self.skip(offset)?;
+        self.read_slice(count)
+    }
+
+    /// Read a null terminated string.
+    ///
+    /// Does not assume any encoding.
+    /// Reads past the null byte, but doesn't return it.
+    #[inline]
+    pub fn read_string(&mut self) -> Option<&'data [u8]> {
```

Comment:
Maybe better named `read_c_string`?

Reply:
I'm hesitant to change to that, because it's starting to be specific enough to imply a `CString` or `CStr`, whereas `string` is generic enough to not necessarily mean `String`, and in the context of object file formats, strings are null terminated.